src/yarp.c

```diff
@@ -3262,21 +3284,90 @@ lex_token_type(yp_parser_t *parser) {
         }
       }
 
+      // If we're in a tilde heredoc, we need to calculate the minimum amount of leading
+      // whitespace so that we can "ignore" this amount of whitespace on all lines in the
+      // heredoc.
+      //
+      // In order to calculate the minimum amount of whitespace, we find the end of the
+      // heredoc, and then iterate from the start until the end, counting the whitespace
+      // on every line and using the `ignored_whitespace` field on the heredoc struct to
+      // track the minimum whitespace we see.
+      if (heredoc_struct.indent == YP_HEREDOC_INDENT_TILDE &&
+              heredoc_struct.ignored_whitespace == -1) {
+        int min_non_newline_whitespace = heredoc_struct.ignored_whitespace;
+
+        const char *end_of_heredoc = parser->current.end - 1;
+
+        // Increment the end_of_heredoc pointer until we find something that matches the
+        // start
+        while(strncmp(end_of_heredoc, ident_start, ident_length) != 0) {
+            end_of_heredoc++;
+            if (end_of_heredoc >= parser->end) {
```

Comment:
This is the case where the heredoc isn't properly closed. I put this line in temporarily so it doesn't segfault, before we decide what do to here.

Mirror ripper and lex all lines as strings?